### PR TITLE
Enhancing netCDF dialog

### PR DESCRIPTION
**Pull request by stevekogo**

### CODE REVIEW

**Comment by rdstern** [comment]
Delighted you are enhancing this command.  I was looking at the ncdf4 facilities and trying to think what users may want.  I notice there is an ncatt_get function.  Given files might be quite large, I wonder if there is scope for:
a) Being able to find out about the file, without reading the data necessarily, e.g. what element, what units, etc.  This could perhaps be an option at the top of the dialogue, i.e,. Provide Details and Read Data could be 2 options.
b) Then would it be easy to read just a part of the file, or would it usually be better to read it all and then use our filters?

I wonder if you and Danny could discuss these options.  They are not something to just put in - it needs discussion beforehand.

**Comment by dannyparsons** [comment]
I think the selector and receivers for the dimension variables is the most needed first step on this.

@stevekogo is this done now? If so I think it's worth merging since other additions are probably not as urgent as other tasks.

File: instat/dlgOpenNetCDF.vb [hunk 1]
```diff
@@ -74,16 +74,31 @@ Public Class dlgOpenNetCDF
...
+        ucrReceiverLonName.SetParameterIsString()
+        ucrReceiverLonName.Selector = ucrSelectorNetCDF
+        ucrReceiverLonName.SetItemType("nc_dim_variables")
+        ucrReceiverLonName.strNcFilePath = ucrInputFilePath.GetText()
+
+        ucrReceiverTimeName.SetParameter(New RParameter("time_col_name", 5))
+        ucrReceiverTimeName.SetParameterIsString()
+        ucrReceiverTimeName.Selector = ucrSelectorNetCDF
+        ucrReceiverTimeName.SetItemType("nc_dim_variables")
+        ucrReceiverTimeName.strNcFilePath = ucrInputFilePath.GetText()
```
Comment (dannyparsons) [inline instat/dlgOpenNetCDF.vb]: You only need to set the path when it is changed. At this point on initialise it's empty anyway.

**Comment by stevekogo** [comment]
I got side tracked by students but I will tidy this up tonight.

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers This is ready for review.

**Comment by stevekogo** [comment]
**Delighted you are enhancing this command. I was looking at the ncdf4 facilities and trying to think what users may want. I notice there is an ncatt_get function. Given files might be quite large, I wonder if there is scope for:
a) Being able to find out about the file, without reading the data necessarily, e.g. what element, what units, etc. This could perhaps be an option at the top of the dialogue, i.e,. Provide Details and Read Data could be 2 options.
b) Then would it be easy to read just a part of the file, or would it usually be better to read it all and then use our filters?

I wonder if you and Danny could discuss these options. They are not something to just put in - it needs discussion beforehand.**

@rdstern I had a discussion with @dannyparsons and we thought this could be implemented later. I am going to post it as an issue.

**Comment by stevekogo** [comment]
@rdstern I have added date conversion capacity to the code and a checkbox to control this in the dialog. @africanmathsinitiative/developers can someone test this and see how it works.

**Comment by rdstern** [comment]
I am not sure whether Steve has provided data for testing?  Here is a set from Rwanda.
[ENACTS data NetCDF_GrADSdata.zip](https://github.com/africanmathsinitiative/R-Instat/files/1018210/ENACTS.data.NetCDF_GrADSdata.zip)

This was able to be read by Steve's dialogue before, but the first column did not appear as dates.  This should now be what has changed.

It is a large file - 3 million observations!  But not many columns.

**Comment by stevekogo** [comment]
@rdstern In the implementation, I preserved the initial numeric column but there is an option for additional date/time column.

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers Is anyone testing this? It should be in good form for @dannyparsons

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers Once this is merged, you will need to have netcdf.tools package installed in order to compile the solution

**Comment by lilyclements** [comment]
@stevekogo are you going to merge the changes in #3085 out of interest?
I'll review after that's confirmed!

**Comment by stevekogo** [comment]
@dannyparsons I will revert this pull request and post is as an issue

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers This is ready for final reviews

**Comment by lilyclements** [comment]
@stevekogo reviewing/making changes to this now

**Comment by lilyclements** [comment]
@stevekogo I have made changes to this dialog. I haven't been able to do a thorough review yet, however, since I couldn't run it. This is to do with removing the packages and it not being called in the instat_object. I believe @dannyparsons is looking into it now. Once that is fixed, I'll continue my review!

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers Seems there is a problem with this. I will fix it in due course.

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers This is now fixed. Someone can have a quick look before @dannyparsons looks at it.

**Comment by lilyclements** [comment]
1. When we open this dialog, it opens straight to the "Open File" and then you press OK and are at the dialog.
This is fine, but other importing dialogs don't do this. I was wondering why this one does?

2. I got an error "There is no package called `ncdf.tools`" since I didn't have that package installed on my computer. Do we need to add that to the installed packages for when downloading R-Instat?

3. After fixing this, to one of my datasets (which I can send to you if you want) I got an error that "Unit months is not implemented.". What does this mean? It would only occur when the check box was checked

4. By default "lat" went into my "Time" selector. Or another time "Longitude" went in it.

image

image

I cannot work out why this is by looking at the code.

5. If I inputted a variable into the receiver which wasn't one of the specified KVPs for that receiver, (e.g. I'd put X in the Y and Y in the X), I'd get the following error:

>   Latitude and longitude names could not be recognised.

File: instat/dlgOpenNetCDF.vb [hunk 2]
```diff
@@ -55,53 +58,93 @@ Public Class dlgOpenNetCDF
...
         ucrInputFilePath.SetParameter(New RParameter("filename", 0))
         ucrInputDataName.SetParameter(New RParameter("main_data_name", 1))
         ucrInputLocDataName.SetParameter(New RParameter("loc_data_name", 2))
-        ucrInputLatColName.SetParameter(New RParameter("latitude_col_name", 3))
-        ucrInputLatColName.SetLinkedDisplayControl(lblLatColName)
-        ucrInputLonColName.SetParameter(New RParameter("longitude_col_name", 4))
-        ucrInputLonColName.SetLinkedDisplayControl(lblLonColName)
+
+
+
```
Comment (lilyclements) [inline instat/dlgOpenNetCDF.vb]: Also there's three blank lines here, can you reduce this to one?

**Comment by stevekogo** [comment]
@lilyclements Can you send me the chirps data so that I can try it out.

**Comment by lilyclements** [comment]
I can send the Chirps one. However, it takes a long time to download on this dialog. I would recommend sending ncdata which is having the issues outlined on 3. and 4.
I'll email it to you now.

**Comment by stevekogo** [comment]
@lilyclements File>Open from File opens the same way I this dialog does. I welcome reactions from @dannyparsons  on how this dialog should open. I will take action once this is clarified.

File: instat/dlgOpenNetCDF.vb (hunk 2 above)
Comment (stevekogo) [inline instat/dlgOpenNetCDF.vb]: I don't understand the cause of issue 3 but issue 4 is fixed.
As for issue 5, that is what is expected, the user should do something sensible else an error should be expected.

**Comment by dannyparsons** [comment]
Yes, happy with the behaviour of the dialog for opening files.

**Comment by lilyclements** [comment]
1+5: Great, thanks for confirming these are as expected

I don't understand 3 either. I'll test these NetCDF changes on an older branch to see if its from a bug introduced in removing the packages or something.

Glad 4 is fixed. 

Issue 2: @dannyparsons do we add ncdf.tools to the list of packages when installing R-Instat (not when opening R-Instat!). I didn't have this installed on my computer, so I don't know if this is an issue that others will have.

**Comment by dannyparsons** [comment]
There's an open issue about that package so it should get added for the next release.

**Comment by dannyparsons** [comment]
Anything still needing to be addressed here?

**Comment by stevekogo** [comment]
This is in god shape for you.

**Comment by rdstern** [comment]
I'm not sure.  I know what I would like from this importing, which is primarily to be able to import time series data for a limited set of points, which we can specify in a number of ways.  I do not have enough details of using the current dialogue to assess whether this is therefore finished.  Do we (can we) have some examples?

**Comment by stevekogo** [comment]
I didn't manage to limit the number of points to import. I am not aware of the functions that can help us do so.

**Comment by stevekogo** [comment]
@rdstern I have done some research and found that there is a way we can subset a netcdf file. This is done through a function netcdf.tools::transNcdfSubset. I will try to find time to implement this if possible, else @dannyparsons Might have to merge what is already implemented and I will revisit this later in the week.

**Comment by stevekogo** [comment]
Also, there is an option of merging netcdf files...this could be done as a separate dialog probably with the subsetting. I will add this to the issues page in a short while.

**Comment by stevekogo** [comment]
@africanmathsinitiative/developers I think this is in good shape to go in, in case it doesn't someone else should  take over.

**Comment by lilyclements** [comment]
Taking over for this

**Comment by lilyclements** [comment]
@dannyparsons this can be closed because I have merged these changes into #3395

**Pull request closed without merging**